cli/command/cli.go

```diff
@@ -145,13 +147,38 @@ func getConfiguredCredentialStore(c *configfile.ConfigFile, serverAddress string
 	return c.CredentialsStore
 }
 
+func newAPIClientWithTLSKeyPass(opts *cliflags.CommonOptions, configFile *configfile.ConfigFile, passRetriever notary.PassRetriever) (client.APIClient, error) {
+	client, err := NewAPIClientFromFlags(opts, configFile)
+	if err == nil {
+		return client, nil
+	}
+
+	var (
+		passwd string
+		giveup bool
+	)
+	for attempts := 0; tlsconfig.IsErrEncryptedKey(err); attempts++ {
+		// some code and comments borrowed from notary/trustmanager/keystore.go
+		passwd, giveup, err = passRetriever("private", "encrypted TLS private", false, attempts)
+		// Check if the passphrase retriever got an error or if it is telling us to give up
+		if giveup || err != nil {
+			return nil, fmt.Errorf("private key is encrypted, but could not get passphrase: %v", err)
+		}
+
+		opts.TLSOptions.Passphrase = passwd
+		client, err = NewAPIClientFromFlags(opts, configFile)
+	}
+	return client, err
+}
+
 // Initialize the dockerCli runs initialization that must happen after command
 // line flags are parsed.
 func (cli *DockerCli) Initialize(opts *cliflags.ClientOptions) error {
 	cli.configFile = LoadDefaultConfigFile(cli.err)
 
 	var err error
-	cli.client, err = NewAPIClientFromFlags(opts.Common, cli.configFile)
+	passRetriever := getPassphraseRetriever(cli, make(map[string]string))
```

Comment:
I've just been double checking the notary code for the passphrase retrievers and it's fine to pass `nil` for this map value. This map is a translation layer that takes a possibly internal password identifier/name and converts it to a user friendly representation. It never gets written, only read, and it's perfectly valid to index into a `nil` map, you'll get get back the zero value and `false`.